### ISSUE TITLE
`abstract`

### ISSUE THREAD

**Issue by andy-ms** [JamesHenry/tsep-babylon-test#7]
### Code

```ts
abstract class C {
    abstract m(): void;
}
```

### Eslint

On the class:

```json
"type": "TSAbstractClassDeclaration",
```

On the method:

```json
"type": "TSAbstractMethodDefinition",
```

### Babylon

On the class:

```json
"abstract": true,
...
"type": "ClassDeclaration",
```

On the method:

```json
"abstract": true,
...
"type": "MethodDefinition",
```

Babylon uses existing node kinds with flags added. Eslint uses a new node kind.
I think the idea with eslint is that rules will crash if they run into a method with no body -- so by using a new type, backwards-compatibility is preserved. Although we already have methods with no body in the case of overloads.

**Comment by JamesHenry** [comment]
That's exactly right.

@soda0289 has given a lot of thought to this particular problem recently, so tagging him in 😄

**Comment by soda0289** [comment]
I think we should change abstract classes to use the property `abstract` instead of changing the entire class declaration type.

The empty body function are the real problem since the estree spec states that `FunctionExpressions` must have a body. Many ESLint rules depend on this fact.

If we can create new node types for the different types of empty body functions that should be enough to allow all the eslint rules to work correctly. This means not creating a `FunctionExpression` node but instead create `FunctionOverloadDeclaration` and `AbstractMethodDefinition`. I have some ideas in this issue:
https://github.com/eslint/typescript-eslint-parser/issues/253

**Comment by andy-ms** [comment]
I think we should have a single kind of empty function -- this may be a function in a `.d.ts` file *or* an overload *or* a method. Maybe call it `DeclareFunction`.
Note that in babylon a ClassMethod directly has parameters and a body, while in estree a ClassMethod *has* a function. So on the babylon side I'd have `DeclareMethod` and the `estree` plugin would have to turn it into a normal `ClassMethod` with a `DeclareFunction` inside.

**Comment by soda0289** [comment]
I like the idea of `DeclareFunction`/`DeclareMethod` for all empty body functions. The `estree` plugin might want to just keep the `DeclareMethod` instead of creating a `MethodDefinition` with a `DeclarationFunction` since many eslint rules make assumptions that `MethodDefintion` always have function expression. Either way labeling empty functions with their own node type will help.

**Comment by andy-ms** [comment]
But the `FunctionExpression` child is necessary to store the parameters.

**Comment by soda0289** [comment]
Yes `DeclareMethod` would store function parameters. To best handle all rules in eslint `MethodDefinition` node must have a `FunctionExpression` and cannot have a child node without a body. We can always modify the eslint rules but to meet current expectations this would be ideal.
